File: sdk/python/tempmail_sdk/providers/tempgmailer.py
"""
TempGmailer 渠道实现
网站: https://tempgmailer.com

流程：
  1. GET 首页获取 CSRF token 和 Laravel session cookie
  2. POST /get-gmail 创建临时 Gmail 邮箱
  3. POST /get-inbox 获取收件箱
"""

import json
import re
from typing import List, Optional

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "tempgmailer"
BASE_URL = "https://tempgmailer.com"

_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
        "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
    "Accept": "application/json, text/plain, */*",
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
    "Referer": f"{BASE_URL}/",
    "X-Requested-With": "XMLHttpRequest",
    "X-TempGmailer-Auth": "frontend",
    "Content-Type": "application/json",
}


def _init_session() -> tuple:
    """
    访问首页获取 CSRF token 和 session cookie

    Returns:
        (csrf_token, cookie_str)
    """
    import requests

    session = requests.Session()
    session.headers.update({"User-Agent": _HEADERS["User-Agent"]})
    resp = session.get(BASE_URL, timeout=15)
    resp.raise_for_status()

    # 从 HTML 中提取 CSRF token
    match = re.search(
        r'<meta\s+name="csrf-token"\s+content="([^"]+)"', resp.text
    )
    if not match:
        raise RuntimeError("tempgmailer: 无法提取 CSRF token")
    csrf_token = match.group(1)

    # 组装 cookie 字符串
    cookie_str = "; ".join([f"{k}={v}" for k, v in session.cookies.items()])
    if not cookie_str:
        raise RuntimeError("tempgmailer: 未获取到 session cookie")

    return csrf_token, cookie_str


def generate_email(
    duration: int = 30, domain: Optional[str] = None
) -> EmailInfo:
    """
    创建 TempGmailer 临时邮箱

    Args:
        duration: 有效期（分钟），本渠道不使用此参数
        domain: 域名筛选，本渠道不使用此参数

    Returns:
        EmailInfo 对象
    """
    csrf_token, cookies = _init_session()

    headers = {
        **_HEADERS,
        "X-CSRF-TOKEN": csrf_token,
        "Cookie": cookies,
    }

    resp = tm_http.post(
        f"{BASE_URL}/get-gmail",
        headers=headers,
        json={"refresh": True, "adblock": 0},
    )
    resp.raise_for_status()
    data = resp.json()

    if not data.get("success"):
        raise RuntimeError(
            f"tempgmailer: 创建邮箱失败: {data.get('message', '未知错误')}"
        )

    email = (data.get("data") or {}).get("email", "").strip()
    if not email:
        raise RuntimeError("tempgmailer: 创建邮箱失败, 未返回邮箱地址")

    # 将 CSRF token 和 cookie 组合为 JSON 字符串作为 token
    token_payload = json.dumps({"csrfToken": csrf_token, "cookies": cookies})

    return EmailInfo(
        channel=CHANNEL,
        email=email,
        _token=token_payload,
    )


def get_emails(token: str, email: str = "", **kwargs) -> List[Email]:
    """
    获取 TempGmailer 收件箱邮件

    Args:
        token: JSON 格式的认证信息（csrf_token + cookies）
        email: 邮箱地址

    Returns:
        标准化邮件列表
    """
    session = json.loads(token)
    csrf_token = session["csrfToken"]
    cookies = session["cookies"]

    headers = {
        **_HEADERS,
        "X-CSRF-TOKEN": csrf_token,
        "Cookie": cookies,
    }

    resp = tm_http.post(
        f"{BASE_URL}/get-inbox",
        headers=headers,
        json={"email": email, "adblock": 0},
    )
    resp.raise_for_status()
    data = resp.json()

    if not data.get("success"):
        return []

    messages = (data.get("data") or {}).get("messages", [])
    if not isinstance(messages, list):
        return []

    emails: List[Email] = []
    for msg in messages:
        if not isinstance(msg, dict):
            continue

        # from 字段可能是对象 {"address": "..."} 或字符串
        from_field = msg.get("from", "")
        if isinstance(from_field, dict):
            from_field = from_field.get("address", "")

        # body 为完整 HTML，intro 为纯文本摘要
        html_content = msg.get("body") or msg.get("intro", "")
        text_content = msg.get("intro", "")

        raw = {
            "id": msg.get("id", ""),
            "from": from_field,
            "to": email,
            "subject": msg.get("subject", ""),
            "text": text_content,
            "html": html_content,
            "date": msg.get("createdAt", ""),
        }
        emails.append(normalize_email(raw, email))

    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/tempgmailer.rb
# frozen_string_literal: true

require "json"

module TempmailSdk
  module Providers
    # TempGmailer 渠道 — https://tempgmailer.com
    # Laravel + CSRF，POST /get-gmail 创建，POST /get-inbox 获取邮件
    module Tempgmailer
      CHANNEL = "tempgmailer"
      BASE_URL = "https://tempgmailer.com"

      # 从 HTML 中提取 CSRF token
      CSRF_RE = /<meta\s+name="csrf-token"\s+content="([^"]+)"/

      HEADERS = {
        "User-Agent" => "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0",
        "Accept" => "application/json, text/plain, */*",
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8",
        "Referer" => "#{BASE_URL}/",
        "X-Requested-With" => "XMLHttpRequest",
        "X-TempGmailer-Auth" => "frontend",
        "Content-Type" => "application/json"
      }.freeze

      module_function

      def extract_cookies(resp)
        cookies = {}
        resp.set_cookies.each do |line|
          pair = line.split(";").first.to_s.strip
          k, v = pair.split("=", 2)
          cookies[k] = v if k && v
        end
        cookies
      end

      def init_session
        resp = Http.get(BASE_URL, headers: {
          "User-Agent" => HEADERS["User-Agent"],
          "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"
        }, timeout: 15)
        resp.raise_for_status

        m = resp.body.match(CSRF_RE)
        raise "tempgmailer: 无法提取 CSRF token" unless m

        csrf_token = m[1]
        cookies = extract_cookies(resp)
        cookie_str = cookies.map { |k, v| "#{k}=#{v}" }.join("; ")
        raise "tempgmailer: 未获取到 session cookie" if cookie_str.empty?

        [csrf_token, cookie_str]
      end

      # 创建临时邮箱
      # @return [EmailInfo]
      def generate_email
        csrf_token, cookie_str = init_session

        headers = HEADERS.merge(
          "X-CSRF-TOKEN" => csrf_token,
          "Cookie" => cookie_str
        )

        resp = Http.post("#{BASE_URL}/get-gmail", headers: headers,
                                                   json: { "refresh" => true, "adblock" => 0 })
        resp.raise_for_status
        data = resp.json
        raise "tempgmailer: 创建邮箱失败: #{data['message'] || '未知错误'}" unless data["success"]

        email = ((data["data"] || {})["email"] || "").strip
        raise "tempgmailer: 创建邮箱失败，未返回邮箱地址" if email.empty?

        token_payload = JSON.generate({ "csrfToken" => csrf_token, "cookies" => cookie_str })
        EmailInfo.new(channel: CHANNEL, email: email, token: token_payload)
      end

      # 获取邮件列表
      # @param token [String] JSON 格式认证信息
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        session = JSON.parse(token)
        csrf_token = session["csrfToken"]
        cookie_str = session["cookies"]

        headers = HEADERS.merge(
          "X-CSRF-TOKEN" => csrf_token,
          "Cookie" => cookie_str
        )

        resp = Http.post("#{BASE_URL}/get-inbox", headers: headers,
                                                   json: { "email" => email, "adblock" => 0 })
        resp.raise_for_status
        data = resp.json
        return [] unless data["success"]

        messages = (data["data"] || {})["messages"] || []
        return [] unless messages.is_a?(Array)

        messages.filter_map do |msg|
          next unless msg.is_a?(Hash)

          from_field = msg["from"]
          from_field = from_field["address"] if from_field.is_a?(Hash)
          from_field = from_field.to_s

          html_content = msg["body"] || msg["intro"] || ""
          text_content = msg["intro"] || ""

          raw = {
            "id" => msg["id"] || "",
            "from" => from_field,
            "to" => email,
            "subject" => msg["subject"] || "",
            "text" => text_content,
            "html" => html_content,
            "date" => msg["createdAt"] || ""
          }
          Normalize.normalize_email(raw, email)
        end
      end
    end
  end
end
